src/persist-client/src/stats.rs

```diff
@@ -0,0 +1,125 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Aggregate statistics about data stored in persist.
+
+use std::collections::BTreeMap;
+
+use mz_persist::indexed::columnar::ColumnarRecords;
+use mz_persist_types::columnar::{ColumnFormat, PartEncoder, Schema};
+use mz_persist_types::part::{Part, PartBuilder};
+use mz_persist_types::Codec;
+
+use crate::internal::encoding::Schemas;
+
+/// Logic for filtering parts in a read as uninteresting purely from stats about
+/// the data held in them.
+pub trait PartFilter<K, V>: 'static {
+    /// True if the part should be fetched, false if it should be skipped.
+    fn should_fetch(&mut self, stats: &PartStats) -> bool;
+}
+
+/// An implementation of [PartFilter] that always returns true.
+#[derive(Debug)]
+pub struct FetchAllPartFilter;
+
+impl<K, V> PartFilter<K, V> for FetchAllPartFilter {
+    fn should_fetch(&mut self, _stats: &PartStats) -> bool {
+        true
+    }
+}
+
+/// Aggregate statistics about data contained in a [Part].
+///
+/// TODO(mfp): PartialEq, Eq, PartialOrd, Ord are sus.
+#[derive(Debug, Default, PartialEq, Eq, PartialOrd, Ord)]
+pub struct PartStats {
+    // TODO(mfp): Only contains Option<u64> columns in this initial skeleton.
+    pub(crate) key_col_min_max_nulls: BTreeMap<String, (u64, u64, usize)>,
+}
+
+impl PartStats {
+    pub(crate) fn new<K: Codec, V: Codec>(
+        schemas: &Schemas<K, V>,
+        part: &Part,
+    ) -> Result<Self, String> {
+        // TODO(mfp): This impl is all entirely just a placeholder. Zero chance
+        // it works like this by the time it ships.
+        let mut key_cols = part.key_ref();
+        let mut key_col_min_max_nulls = BTreeMap::new();
+        for (name, typ) in schemas.key.columns() {
```

Comment:
It's not clear to me why we need a schema for this bit... the part itself is aware of the column names and types. Can we drop the parameter? (Generally it seems to me that the less code needs to be schema-aware the better, since there are contexts where `Schema`s aren't as readily available.)

Reply:
I have it because I think at some point, we'll need to allow the Schema to provide arbitrary logic for min and max for complex fields (in particular, unions and structs) and I was thinking that it's nice to do all the plumbing at once. I'll leave revisiting this decision as a TODO!